### PR TITLE
[IRGen] Other instructions whose ops/results have packs may allocate pack metadata.

### PR DESCRIPTION
**Pull request by nate-chandler**
Based on https://github.com/apple/swift/pull/67567 .

Instead of assuming that the list of instructions known to allocate pack metadata is exhaustive and returning false from mayRequirePackMetadata for all others, consider the types of the results and operands of other instructions and look for packs.

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test linux platform

File: lib/AST/Type.cpp [hunk 1]
```diff
@@ -3703,8 +3703,9 @@ PackArchetypeType::PackArchetypeType(
     ArrayRef<ProtocolDecl *> ConformsTo, Type Superclass,
     LayoutConstraint Layout, PackShape Shape)
     : ArchetypeType(TypeKind::PackArchetype, Ctx,
-                    RecursiveTypeProperties::HasArchetype, InterfaceType,
-                    ConformsTo, Superclass, Layout, GenericEnv) {
+                    RecursiveTypeProperties::HasArchetype |
+                        RecursiveTypeProperties::HasConcretePack,
```
Comment (slavapestov) [inline lib/AST/Type.cpp:3707]: It’s actually PackExpansionType that leads us into this situation

Comment (nate-chandler) [reply lib/AST/Type.cpp:3707]: Yeah, it's a PackExpansionType whose pattern is a PackArchetypeType:

```
(tuple_type num_elements=1
  (tuple_type_elt
    (pack_expansion_type
      (pattern=pack_archetype_type address=0x127045b10
        (interface_type=dependent_member_type assoc_type=Swift.(file).Sequence.Element
          (base=generic_type_param_type depth=0 index=0 decl=main.(file).tupleExpansionWithMemberType(seqs:elts:).T@main.swift:1:47 pack)) name=(each T).Element
)
      (count=pack_archetype_type address=0x127045880
        (interface_type=generic_type_param_type depth=0 index=0 decl=main.(file).tupleExpansionWithMemberType(seqs:elts:).T@main.swift:1:47 pack) conforms_to=Swift.(file).Sequence name=each T
))))
```

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by NRC1978** [comment]
The beach

**Pull request merged**